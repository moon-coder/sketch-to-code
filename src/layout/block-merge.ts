import {INode} from "../types";

/**
 * @desc
 * 布局第二步,行合并算法
 * 参考https://www.yuque.com/dscga7/bz0gzn/neh95s#PDMkL
 *
 * @使用场景
 *
 * @coder.yang2010@gmail.com
 * @Date    2019/10/15
 **/
export default function (nodes: INode[]):INode{


  return nodes[0];
}