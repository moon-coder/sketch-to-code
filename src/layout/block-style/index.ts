import colStyle from './col-style';
import rowStyle from './row-style';
import {INode} from "../../types";

export default function (node: INode) {
  colStyle(node);
  rowStyle(node);
  return node;
}
