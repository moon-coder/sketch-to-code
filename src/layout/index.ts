import { INode } from "../types";
import {
  isOverlap,
  calcBoundaryNode,
  isContainer,
  calcNodeCoordsNew,
  calcInnerSpacings,
  calcBoundaryBox,
  calcHSpacing,
  mergeLog,
  canMergeLog, walk, sortNodesByLayout
} from "./utils";
import rowMerge from  './row-merge';
import blockMerge from  './block-merge';

/**
 *
 */
const phaseOne = (nodes: INode[]) => {
  // 1.预处理  包含关系的划分;

  const rootNode = preDeal(nodes);


  walk(rootNode,(node) => {
    if(node.children && node.children.length >0) {
      if(node.children.length > 1) {
        node.children = rowMerge(node.children);
      }
      if(node.children.length > 1) {
        //一个节点只有一个子节点, 则这个关系 可以去除
        node.children = [blockMerge(node.children)];
      }
      return ;
    }
  })
  // 3.去掉同行同列
  mergeLineRow(rootNode);

  // 4.重新排列
  reSort(rootNode);

  // 5.重命名样式
  renameClassName(rootNode);

  if (rootNode.children.length == 1) return rootNode.children[0];

  return rootNode;
}

/**
 * 预处理 image text 等叶子 节点倒推一步;
 */
const preDeal = (nodes: INode[]): INode => {
  nodes  =nodes.sort((a,b)=>a.frame.y-b.frame.y);

  const root = calcBoundaryNode(nodes);

  const children: INode[] = root.children;
  children.forEach(node => node.points = calcNodeCoordsNew(node));


  root.children = [];

  // 将结点加入最近的外框结点
  const addToOuter = (node: INode, nodes: INode[]) => {
    const outers = nodes.filter(item => item.type === 'Block' && item != node && isContainer(node, item));
    outers.sort((a, b) =>  isContainer(a, b) ? -1 : 1);
    outers[0].children.push(node);
  }
  // 遍历所有节点，将其加入最近外框
  children.forEach(child => {
    addToOuter(child, [root, ...children])
  });
  return root;
}

/**
 * 合并所有确定结点
 */
const mergeSure = (node: INode) => {

  // 1.walk整棵布局树
  const walkNode = (node: INode): boolean => {
    if (node.type === 'Block' && !node.style.flexDirection) {
      const childrenFlag = node.children.every(child => walkNode(child));
      // 2.针对Block结点进行确定合并
      const nodeFlag = mergeNode(node);
      // 利用返回参数 fixme
      return childrenFlag && nodeFlag;
    }
    return true;
  }
  return walkNode(node);
}

/**
 * 2.对Block 下children 结点逐个进行合并测试
 *  两个节点
 *
 * @param {INode} node
 * @returns {boolean}
 */
function mergeNode(node: INode):boolean {


  if (node.children.length == 0) return true;
  let k = 0;
  // 2.1.不断循环遍历node的子结点
  while (true) {
    k++; if(k > 20) throw new Error('超出循环上线2');
    for (let i = 0; i < node.children.length; i++) {

      //选中对比节点
      const child = node.children[i];
      const others = node.children.filter(item => item != child);

      const canMerges =  others.filter(otherNode => {
        //otherNode =>即将合并测试节点

        //extras 其他节点
        const extras = others.filter(item => item != otherNode);

        return extras.every(extra => !isOverlap(calcBoundaryNode([child, otherNode]), extra));
      });
      canMergeLog(child, canMerges);
      if (canMerges.length == 1) {
        // 2.2.当发现可以进行一次确定的合并时，合并并重新遍历；
        console.log('进行了一次确定合并：' + mergeLog(child, canMerges[0]));
        // 判断方向
        const boundaryNode = calcBoundaryNode([child, canMerges[0]]);
        if (child.points[1].x <= canMerges[0].points[0].x
          || child.points[0].x >= canMerges[0].points[1].x) {
          boundaryNode.style.flexDirection = 'row';
        } else {
          boundaryNode.style.flexDirection = 'column';
        }
        // 合并结点
        node.children = node.children
          .filter(item => !(item == child || item == canMerges[0]))
          .concat(boundaryNode);
        break;
      }

      if (i == (node.children.length - 1)) {
        // 2.3.node子结点遍历完成后返回
        // fixme 能不能这么判断
        return node.children.length == 1 || node.children.length == 0;
      }
    }
  }


}

/**
 * 合并一次可选结点
 */
const mergeOptional = (node: INode) => {
  const mergeNode = (node: INode) => {
    for (let i = 0; i < node.children.length; i++) {
      const child = node.children[i];
      const others = node.children.filter(item => item != child);
      const canMerges =  others.filter(other => {
        const extras = others.filter(item => item != other);
        return extras.every(extra => !isOverlap(calcBoundaryNode([child, other]), extra));
      });
      canMergeLog(child, canMerges);
      console.log(`${node.children.length}  ${canMerges.length}`);
      if (canMerges.length > 0) { // fixme 这边判断条件还要改
        console.log('进行了一次选择合并：' + mergeLog(child, canMerges[0]));
        const boundaryNode = calcBoundaryNode([child, canMerges[0]]);
        if (child.points[1].x <= canMerges[0].points[0].x
            || child.points[0].x >= canMerges[0].points[1].x) {
          boundaryNode.style.flexDirection = 'row';
        } else {
          boundaryNode.style.flexDirection = 'column';
        }
        // 合并结点
        node.children = node.children
            .filter(item => !(item == child || item == canMerges[0]))
            .concat(boundaryNode);
        return;
      }
    }
  }

  // 1.walk整棵布局树
  const walkNode = (node: INode) => {
    if (node.type === 'Block' && !node.style.flexDirection) {
      node.children.every(child => walkNode(child));
      // 2.针对Block结点进行确定合并
      mergeNode(node);
    }
  }
  walkNode(node);
}

/**
 *
 * 如果节点的父与此节点flexDirection 一致, 且此节点只有一个节点, 那么除去此中间节点;
 * TODO 要考虑下这个中间节点 什么时候产生的;
 *
 * @param {INode} node
 */
const mergeLineRow = (node: INode) => {
  const children = [...node.children];


  //建立父子关系
  children.forEach(child => {
    child.parent = node;
    mergeLineRow(child);
  });


  if (node.parent
      && ( (node.parent.style.flexDirection && node.parent.style.flexDirection == node.style.flexDirection )|| node.children.length == 1)) {
    node.parent.children.splice(node.parent.children.indexOf(node), 1, ...node.children);
    node.children.forEach(child => child.parent = node.parent);
  }
}

const reSort = (node: INode) => {
  if (node.type === 'Block' && node.style.flexDirection) {
    node.children.sort((a, b) => {
      if (node.style.flexDirection === 'row') {
        return a.points[4].x - b.points[4].x;
      } else {
        return a.points[4].y - b.points[4].y;
      }
    });
  }
  node.children.forEach(child => reSort(child));
}

const renameClassName = (node: INode) => {
  let imgIdx = 0, blockIdx = 0, textIdx = 0;
  ((node: INode) => {
    const walk = (node: INode) => {
      if (node.type == 'Text') node.attrs.className = 'text' + textIdx++;
      if (node.type == 'Block') node.attrs.className = 'block' + blockIdx++;
      if (node.type == 'Image') node.attrs.className = 'img' + imgIdx++;
      node.children.forEach(child => walk(child));
    }
    walk(node);
  })(node);
}


/**
 * 布局算法核心;
 * https://www.yuque.com/dscga7/bz0gzn/neh95s#PDMkL
 行优先: 一行且块行高一致,优先合为一组;
 从面积最小元素开始选择合并:
 合关后面积应该最小;
 * @param {INode[]} nodes
 * @returns {INode}
 */
export default (nodes: INode[]): INode => {

  let rootNode =phaseOne(nodes);
  debugger;
  calcRowLayout(rootNode);
  calcColLayout(rootNode);
  return rootNode;
}




/**
 * 行布局计算
 */
const calcRowLayout = (node: INode) => {

  node.children.forEach(child => calcRowLayout(child));

  if (node.style.flexDirection !== 'row') return;

  // 子结点是否为等分分布
  const isNDivide = (node: INode) => {
    // 依次判断子结点中点是否在等分点上
    for (let i = 0; i < children.length; i++) {
      const x1 = node.points[0].x, n = children.length, w = node.frame.width;
      let x = x1 + 1 / 2 * n + (i - 1) * w / n;
      if (children[i].points[4].x != x) {
        return false;
      }
    }
    return true;
  }

  const { width: nodeWidth } = node.frame;
  const connSpace = nodeWidth / 10;
  // fixme 要用到breakSpace
  const breakSpace = nodeWidth / 5;
  const children = node.children;
  const boundaryBox = calcBoundaryBox(node.children);

  // 1.计算justifyContent
  if (children.length == 1 && children[0].points[4].x == node.points[4].x) {
    // 1.1.单节点、外框中点，center
    node.style.justifyContent = 'center';
  }

  // 计算leftBoxes和rightBoxes
  let leftBoxes: INode[], rightBoxes: INode[] = [];
  children.sort((a, b) => a.points[4].x - b.points[4].x)
  let rightBoundary = node.points[1].x;
  let rightFlag = false;
  for (let i = children.length - 1; i >= 0; i--) {
    if (rightBoundary - children[i].points[1].x < connSpace) {
      rightBoxes.push(children[i]);
      rightBoundary = children[i].points[0].x;
    } else if(rightBoxes.length > 0) {
      rightFlag = true;
    } else {
      break;
    }
  }
  if (!rightFlag) rightBoxes = [];
  leftBoxes = children.filter(child => !rightBoxes.includes(child));

  if (rightBoxes.length == 0) {
    // 1.2.全部为左结点，flex-start
    node.style.justifyContent = 'flex-start';
    node.style.paddingLeft = calcInnerSpacings(node, boundaryBox)[3];
    if(isNDivide(node)) {
      // 平均分布
    } else {
      // 计算marginRight
      children.forEach((child, idx) => {
        if (idx == 0) return;
        children[idx - 1].style.marginRight = calcHSpacing(children[idx - 1], child);
      });
    }
  } else if(leftBoxes.length == 0) {
    // 1.3.全部为右结点，flex-end
    node.style.justifyContent = 'flex-end';
    node.style.paddingRight = calcInnerSpacings(node, boundaryBox)[1];
    // 计算marginLeft
    children.forEach((child, idx) => {
      if (idx == children.length - 1) return;
      children[idx + 1].style.marginLeft = calcHSpacing(child, children[idx + 1]);
    });
  } else {
    // 1.4.既有左结点又有右结点，space-between
    node.style.justifyContent = 'space-between';
    node.style.paddingLeft = calcInnerSpacings(node, boundaryBox)[3];
    node.style.paddingRight = calcInnerSpacings(node, boundaryBox)[1];
    node.children = [];
    if (leftBoxes.length > 1) {
      const leftBox = calcBoundaryNode(leftBoxes);
      leftBox.style.flexDirection = 'row';
      node.children.push(leftBox);
      // 计算marginRight
      leftBoxes.forEach((box, idx) => {
        if (idx == 0) return;
        leftBoxes[idx - 1].style.marginRight = calcHSpacing(leftBoxes[idx - 1], box);
      });
    } else {
      node.children.push(...leftBoxes);
    }
    if (rightBoxes.length > 1) {
      // 计算marginLeft
      const rightBox = calcBoundaryNode(rightBoxes);
      rightBox.style.flexDirection = 'row';
      node.children.push(rightBox);
      rightBoxes.forEach((box, idx) => {
        if (idx == rightBoxes.length - 1) return;
        rightBoxes[idx + 1].style.marginLeft = calcHSpacing(box, rightBoxes[idx + 1]);
      });
    } else {
      node.children.push(...rightBoxes);
    }
  }

  // 2.计算alignItems
  let alignItemsMap = {
    'center': [],
    'flex-start': [],
    'flex-end': [],
    'stretch': [],
    'others': []
  } as any;
  // 根据坐标位置，计算每个节点的布局位置
  children.forEach((child: INode) => {
    if (child.points[4].y == node.points[4].y) {
      // 2.1.外框中点，center
      alignItemsMap['center'].push(child);
    } else if (child.points[0].y == boundaryBox[0].y && child.points[3].y == boundaryBox[3].y) {
      // 2.2.包围框上下边，stretch
      alignItemsMap['stretch'].push(child);
    } else if(child.points[0].y == boundaryBox[0].y) {
      // 2.3.包围框上边，flex-start
      alignItemsMap['flex-start'].push(child);
    } else if(child.points[3].y == boundaryBox[3].y) {
      // 2.4.包围框下边，flex-end
      alignItemsMap['flex-end'].push(child);
    } else {
      alignItemsMap['others'].push(child);
    }
  });
  // 选取子结点中出现最多的布局作为alignItems的布局，其余非主流布局的结点设置自己的alignSelf
  const alignItemsArr = Object.entries(alignItemsMap).sort((a: any[], b: any[]) => b[1].length - a[1].length);
  const key = alignItemsArr[0][0];
  node.style.alignItems = key;
  // 设置padding
  if(key === 'stretch') {
    node.style.paddingTop = calcInnerSpacings(node, boundaryBox)[0];
    node.style.paddingBottom = calcInnerSpacings(node, boundaryBox)[2];
  } else if(key === 'flex-start') {
    node.style.paddingTop = calcInnerSpacings(node, boundaryBox)[0];
  } else if(key === 'flex-end') {
    node.style.paddingBottom = calcInnerSpacings(node, boundaryBox)[2];
  }
  alignItemsArr.forEach((item: any, idx) => {
    if (idx == 0) return;
    if (item[0] === 'others') {
      // 给alignSelf:flex-start和marginTop
      item[1].forEach((child: any) => {
        child.style.alignSelf = 'flex-start';
        child.style.marginTop = calcInnerSpacings(node, child.points)[0];
      });
    } else {
      item[1].forEach((n: INode) => n.style.alignSelf = item[0]);
    }
  });
}

/**
 * 列布局计算
 */
const calcColLayout = (node: INode) => {

  node.children.forEach(child => calcColLayout(child));

  if (node.style.flexDirection !== 'column') return;

  const children = node.children;
  const boundaryBox = calcBoundaryBox(node.children);

  // 1.justifyContent计算
  if (children.length == 1 && children[0].points[4].y == node.points[4].y) {
    // 1.1.单节点外框中点，center
    node.style.justifyContent = 'center';
  } else {
    // 1.2.其它情况，flex-start
    node.style.justifyContent = 'flex-start';
    node.style.paddingTop = calcInnerSpacings(node, boundaryBox)[0];
  }

  // 2.alignItems计算
  let alignItemsMap = {
    'center': [],
    'flex-start': [],
    'flex-end': [],
    'stretch': [],
    'others': []
  } as any;
  // 根据坐标位置，计算每个节点的布局位置
  children.forEach(child => {
    if (child.points[4].x == node.points[4].x) {
      // 2.1.外框中点，center
      alignItemsMap['center'].push(child);
    } else if (child.points[0].x == boundaryBox[0].x && child.points[1].x == boundaryBox[1].x) {
      // 2.2.包围框左右边，stretch
      alignItemsMap['stretch'].push(child);
    } else if(child.points[0].x == boundaryBox[0].x) {
      // 2.3.包围框左边，flex-start
      alignItemsMap['flex-start'].push(child);
    } else if(child.points[1].x == boundaryBox[1].x) {
      // 2.4.包围框右边，flex-end
      alignItemsMap['flex-end'].push(child);
    } else {
      // 2.5.其它情况
      alignItemsMap['others'].push(child);
    }
  });
  // 选取子结点中出现最多的布局作为alignItems的布局，其余非主流布局的结点设置自己的alignSelf
  const alignItemsArr = Object.entries(alignItemsMap).sort((a: any[], b: any[]) => b[1].length - a[1].length);
  const key = alignItemsArr[0][0];
  // 设置padding
  if(key === 'stretch') {
    node.style.paddingLeft = calcInnerSpacings(node, boundaryBox)[3];
    node.style.paddingRight = calcInnerSpacings(node, boundaryBox)[1];
  } else if(key === 'flex-start') {
    node.style.paddingLeft = calcInnerSpacings(node, boundaryBox)[3];
  } else if(key === 'flex-end') {
    node.style.paddingRight = calcInnerSpacings(node, boundaryBox)[1];
  }
  node.style.alignItems = key;
  alignItemsArr.forEach((item: any, idx) => {
    if (idx == 0) return;
    if (item[0] === 'others') {
      // 给alignSelf:flex-start和marginTop
      item[1].forEach((child: INode) => {
        child.style.alignSelf = 'flex-start';
        child.style.marginLeft = calcInnerSpacings(node, child.points)[3];
      });
    } else {
      item[1].forEach((n: INode) => n.style.alignSelf = item[0]);
    }
  });

}

