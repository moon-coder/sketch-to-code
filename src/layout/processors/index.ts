import {IProcessor} from "../../types";
import sameNodesMargin from './same-nodes-margin';

export default <IProcessor[]> [
  sameNodesMargin
];
