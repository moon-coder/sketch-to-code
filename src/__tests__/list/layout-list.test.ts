import {INode} from "../../types";
import getNodes from "../../get-nodes";
import layout from "../../layout";
import {readJSONSync,writeJSONSync} from  'fs-extra';
import {join} from 'path';

/**
 * @desc
 *
 * @使用场景
 *
 * @coder.yang2010@gmail.com
 * @Date    2019/10/14
 **/


it('简易列表代码生成', function () {
  console.log("hello");
  //TODO 等图片的问题处理好.

  let layers = readJSONSync(join(__dirname,"origin.json"));
  const nodes: INode[] = getNodes(layers.layers[0]);

  // 布局处理
  const node: INode = layout(nodes);

  const result =toJSON(node);

  writeJSONSync(join(__dirname,"list-layout-result.json"),result);
  expect(result).toMatchSnapshot("list-layout");
});


function toJSON(node: INode) {

  return JSON.parse(JSON.stringify(node, (key, value)=>{
    if(key === "parent") {
      return  undefined;
    }
    return value;
  },2))
}