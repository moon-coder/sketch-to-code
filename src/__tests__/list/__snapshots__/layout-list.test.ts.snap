// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list-item : list-item 1`] = `
Object {
  "attrs": Object {
    "className": "block0",
  },
  "children": Array [],
  "frame": Object {
    "height": 80,
    "width": 327,
    "x": 12,
    "y": 48,
  },
  "points": Array [
    Object {
      "x": 12,
      "y": 48,
    },
    Object {
      "x": 339,
      "y": 48,
    },
    Object {
      "x": 339,
      "y": 128,
    },
    Object {
      "x": 12,
      "y": 128,
    },
    Object {
      "x": 175.5,
      "y": 88,
    },
  ],
  "style": Object {},
  "type": "Block",
}
`;

exports[`list-origin1 : list-origin1 1`] = `
Object {
  "attrs": Object {
    "className": "block1",
  },
  "children": Array [
    Object {
      "attrs": Object {
        "className": "text0",
        "text": "为您推荐",
      },
      "children": Array [],
      "frame": Object {
        "height": 16,
        "width": 64,
        "x": 24,
        "y": 420,
      },
      "points": Array [
        Object {
          "x": 24,
          "y": 420,
        },
        Object {
          "x": 88,
          "y": 420,
        },
        Object {
          "x": 88,
          "y": 436,
        },
        Object {
          "x": 24,
          "y": 436,
        },
        Object {
          "x": 56,
          "y": 428,
        },
      ],
      "style": Object {
        "alignSelf": "flex-start",
        "color": "#333333ff",
        "fontSize": 16,
        "fontWeight": 5,
        "letterSpacing": 0,
        "lineHeight": 16,
      },
      "type": "Text",
    },
    Object {
      "attrs": Object {
        "className": "block2",
      },
      "children": Array [
        Object {
          "attrs": Object {
            "className": "block3",
          },
          "children": Array [
            Object {
              "attrs": Object {
                "className": "img0",
                "src": "./img/img0.png",
              },
              "children": Array [],
              "frame": Object {
                "height": 80,
                "width": 80,
                "x": 24,
                "y": 452,
              },
              "points": Array [
                Object {
                  "x": 24,
                  "y": 452,
                },
                Object {
                  "x": 104,
                  "y": 452,
                },
                Object {
                  "x": 104,
                  "y": 532,
                },
                Object {
                  "x": 24,
                  "y": 532,
                },
                Object {
                  "x": 64,
                  "y": 492,
                },
              ],
              "style": Object {
                "alignSelf": "center",
                "marginRight": 116,
              },
              "type": "Image",
            },
            Object {
              "attrs": Object {
                "className": "text1",
                "text": "¥20～¥160",
              },
              "children": Array [],
              "frame": Object {
                "height": 16,
                "width": 81,
                "x": 116,
                "y": 512,
              },
              "points": Array [
                Object {
                  "x": 116,
                  "y": 512,
                },
                Object {
                  "x": 197,
                  "y": 512,
                },
                Object {
                  "x": 197,
                  "y": 528,
                },
                Object {
                  "x": 116,
                  "y": 528,
                },
                Object {
                  "x": 156.5,
                  "y": 520,
                },
              ],
              "style": Object {
                "color": "#ff6600ff",
                "fontSize": 16,
                "fontWeight": 5,
                "letterSpacing": 0,
                "lineHeight": 16,
                "marginRight": 116,
              },
              "type": "Text",
            },
            Object {
              "attrs": Object {
                "className": "text2",
                "text": "大型犬只野兽派洗剪吹，凶悍大叔专业洗狗12年，洗遍天下无敌狗啦啦…",
              },
              "children": Array [],
              "frame": Object {
                "height": 36,
                "width": 235,
                "x": 116,
                "y": 456,
              },
              "points": Array [
                Object {
                  "x": 116,
                  "y": 456,
                },
                Object {
                  "x": 351,
                  "y": 456,
                },
                Object {
                  "x": 351,
                  "y": 492,
                },
                Object {
                  "x": 116,
                  "y": 492,
                },
                Object {
                  "x": 233.5,
                  "y": 474,
                },
              ],
              "style": Object {
                "color": "#333333ff",
                "fontSize": 14,
                "fontWeight": 5,
                "letterSpacing": 0,
                "lineHeight": 18,
              },
              "type": "Text",
            },
          ],
          "frame": Object {
            "height": 80,
            "width": 327,
            "x": 24,
            "y": 452,
          },
          "points": Array [
            Object {
              "x": 24,
              "y": 452,
            },
            Object {
              "x": 351,
              "y": 452,
            },
            Object {
              "x": 351,
              "y": 532,
            },
            Object {
              "x": 24,
              "y": 532,
            },
            Object {
              "x": 187.5,
              "y": 492,
            },
          ],
          "style": Object {
            "alignItems": "others",
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "paddingLeft": 0,
          },
          "type": "Block",
        },
        Object {
          "attrs": Object {
            "className": "block4",
          },
          "children": Array [
            Object {
              "attrs": Object {
                "className": "img1",
                "src": "./img/img2.png",
              },
              "children": Array [],
              "frame": Object {
                "height": 80,
                "width": 80,
                "x": 24,
                "y": 544,
              },
              "points": Array [
                Object {
                  "x": 24,
                  "y": 544,
                },
                Object {
                  "x": 104,
                  "y": 544,
                },
                Object {
                  "x": 104,
                  "y": 624,
                },
                Object {
                  "x": 24,
                  "y": 624,
                },
                Object {
                  "x": 64,
                  "y": 584,
                },
              ],
              "style": Object {
                "alignSelf": "center",
                "marginRight": 116,
              },
              "type": "Image",
            },
            Object {
              "attrs": Object {
                "className": "text3",
                "text": "¥20～¥75",
              },
              "children": Array [],
              "frame": Object {
                "height": 16,
                "width": 73,
                "x": 116,
                "y": 604,
              },
              "points": Array [
                Object {
                  "x": 116,
                  "y": 604,
                },
                Object {
                  "x": 189,
                  "y": 604,
                },
                Object {
                  "x": 189,
                  "y": 620,
                },
                Object {
                  "x": 116,
                  "y": 620,
                },
                Object {
                  "x": 152.5,
                  "y": 612,
                },
              ],
              "style": Object {
                "color": "#ff6600ff",
                "fontSize": 16,
                "fontWeight": 5,
                "letterSpacing": 0,
                "lineHeight": 16,
                "marginRight": 116,
              },
              "type": "Text",
            },
            Object {
              "attrs": Object {
                "className": "text4",
                "text": "大型犬野兽派洗剪吹A套餐",
              },
              "children": Array [],
              "frame": Object {
                "height": 18,
                "width": 235,
                "x": 116,
                "y": 548,
              },
              "points": Array [
                Object {
                  "x": 116,
                  "y": 548,
                },
                Object {
                  "x": 351,
                  "y": 548,
                },
                Object {
                  "x": 351,
                  "y": 566,
                },
                Object {
                  "x": 116,
                  "y": 566,
                },
                Object {
                  "x": 233.5,
                  "y": 557,
                },
              ],
              "style": Object {
                "color": "#333333ff",
                "fontSize": 14,
                "fontWeight": 5,
                "letterSpacing": 0,
                "lineHeight": 18,
              },
              "type": "Text",
            },
          ],
          "frame": Object {
            "height": 80,
            "width": 327,
            "x": 24,
            "y": 544,
          },
          "points": Array [
            Object {
              "x": 24,
              "y": 544,
            },
            Object {
              "x": 351,
              "y": 544,
            },
            Object {
              "x": 351,
              "y": 624,
            },
            Object {
              "x": 24,
              "y": 624,
            },
            Object {
              "x": 187.5,
              "y": 584,
            },
          ],
          "style": Object {
            "alignItems": "others",
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "paddingLeft": 0,
          },
          "type": "Block",
        },
        Object {
          "attrs": Object {
            "className": "block5",
          },
          "children": Array [
            Object {
              "attrs": Object {
                "className": "img2",
                "src": "./img/img1.png",
              },
              "children": Array [],
              "frame": Object {
                "height": 80,
                "width": 80,
                "x": 24,
                "y": 636,
              },
              "points": Array [
                Object {
                  "x": 24,
                  "y": 636,
                },
                Object {
                  "x": 104,
                  "y": 636,
                },
                Object {
                  "x": 104,
                  "y": 716,
                },
                Object {
                  "x": 24,
                  "y": 716,
                },
                Object {
                  "x": 64,
                  "y": 676,
                },
              ],
              "style": Object {
                "alignSelf": "center",
                "marginRight": 116,
              },
              "type": "Image",
            },
            Object {
              "attrs": Object {
                "className": "text5",
                "text": "¥20～¥160",
              },
              "children": Array [],
              "frame": Object {
                "height": 16,
                "width": 81,
                "x": 116,
                "y": 696,
              },
              "points": Array [
                Object {
                  "x": 116,
                  "y": 696,
                },
                Object {
                  "x": 197,
                  "y": 696,
                },
                Object {
                  "x": 197,
                  "y": 712,
                },
                Object {
                  "x": 116,
                  "y": 712,
                },
                Object {
                  "x": 156.5,
                  "y": 704,
                },
              ],
              "style": Object {
                "color": "#ff6600ff",
                "fontSize": 16,
                "fontWeight": 5,
                "letterSpacing": 0,
                "lineHeight": 16,
                "marginRight": 116,
              },
              "type": "Text",
            },
            Object {
              "attrs": Object {
                "className": "text6",
                "text": "大型犬只野兽派洗剪吹，凶悍大叔专业洗狗12年，洗遍天下狗",
              },
              "children": Array [],
              "frame": Object {
                "height": 36,
                "width": 235,
                "x": 116,
                "y": 640,
              },
              "points": Array [
                Object {
                  "x": 116,
                  "y": 640,
                },
                Object {
                  "x": 351,
                  "y": 640,
                },
                Object {
                  "x": 351,
                  "y": 676,
                },
                Object {
                  "x": 116,
                  "y": 676,
                },
                Object {
                  "x": 233.5,
                  "y": 658,
                },
              ],
              "style": Object {
                "color": "#333333ff",
                "fontSize": 14,
                "fontWeight": 5,
                "letterSpacing": 0,
                "lineHeight": 18,
              },
              "type": "Text",
            },
          ],
          "frame": Object {
            "height": 80,
            "width": 327,
            "x": 24,
            "y": 636,
          },
          "points": Array [
            Object {
              "x": 24,
              "y": 636,
            },
            Object {
              "x": 351,
              "y": 636,
            },
            Object {
              "x": 351,
              "y": 716,
            },
            Object {
              "x": 24,
              "y": 716,
            },
            Object {
              "x": 187.5,
              "y": 676,
            },
          ],
          "style": Object {
            "alignItems": "others",
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "paddingLeft": 0,
          },
          "type": "Block",
        },
        Object {
          "attrs": Object {
            "className": "block6",
          },
          "children": Array [
            Object {
              "attrs": Object {
                "className": "img3",
                "src": "./img/img3.png",
              },
              "children": Array [],
              "frame": Object {
                "height": 80,
                "width": 80,
                "x": 24,
                "y": 728,
              },
              "points": Array [
                Object {
                  "x": 24,
                  "y": 728,
                },
                Object {
                  "x": 104,
                  "y": 728,
                },
                Object {
                  "x": 104,
                  "y": 808,
                },
                Object {
                  "x": 24,
                  "y": 808,
                },
                Object {
                  "x": 64,
                  "y": 768,
                },
              ],
              "style": Object {
                "alignSelf": "center",
                "marginRight": 116,
              },
              "type": "Image",
            },
            Object {
              "attrs": Object {
                "className": "text7",
                "text": "¥20～¥75",
              },
              "children": Array [],
              "frame": Object {
                "height": 16,
                "width": 73,
                "x": 116,
                "y": 788,
              },
              "points": Array [
                Object {
                  "x": 116,
                  "y": 788,
                },
                Object {
                  "x": 189,
                  "y": 788,
                },
                Object {
                  "x": 189,
                  "y": 804,
                },
                Object {
                  "x": 116,
                  "y": 804,
                },
                Object {
                  "x": 152.5,
                  "y": 796,
                },
              ],
              "style": Object {
                "color": "#ff6600ff",
                "fontSize": 16,
                "fontWeight": 5,
                "letterSpacing": 0,
                "lineHeight": 16,
                "marginRight": 116,
              },
              "type": "Text",
            },
            Object {
              "attrs": Object {
                "className": "text8",
                "text": "小型犬洗剪吹B套餐",
              },
              "children": Array [],
              "frame": Object {
                "height": 18,
                "width": 235,
                "x": 116,
                "y": 732,
              },
              "points": Array [
                Object {
                  "x": 116,
                  "y": 732,
                },
                Object {
                  "x": 351,
                  "y": 732,
                },
                Object {
                  "x": 351,
                  "y": 750,
                },
                Object {
                  "x": 116,
                  "y": 750,
                },
                Object {
                  "x": 233.5,
                  "y": 741,
                },
              ],
              "style": Object {
                "color": "#333333ff",
                "fontSize": 14,
                "fontWeight": 5,
                "letterSpacing": 0,
                "lineHeight": 18,
              },
              "type": "Text",
            },
          ],
          "frame": Object {
            "height": 80,
            "width": 327,
            "x": 24,
            "y": 728,
          },
          "points": Array [
            Object {
              "x": 24,
              "y": 728,
            },
            Object {
              "x": 351,
              "y": 728,
            },
            Object {
              "x": 351,
              "y": 808,
            },
            Object {
              "x": 24,
              "y": 808,
            },
            Object {
              "x": 187.5,
              "y": 768,
            },
          ],
          "style": Object {
            "alignItems": "others",
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "paddingLeft": 0,
          },
          "type": "Block",
        },
        Object {
          "attrs": Object {
            "className": "text9",
            "text": "更多服务/商品",
          },
          "children": Array [],
          "frame": Object {
            "height": 14,
            "width": 91,
            "x": 142.5,
            "y": 831,
          },
          "points": Array [
            Object {
              "x": 142.5,
              "y": 831,
            },
            Object {
              "x": 233.5,
              "y": 831,
            },
            Object {
              "x": 233.5,
              "y": 845,
            },
            Object {
              "x": 142.5,
              "y": 845,
            },
            Object {
              "x": 188,
              "y": 838,
            },
          ],
          "style": Object {
            "alignSelf": "flex-start",
            "color": "#ff6600ff",
            "fontSize": 14,
            "fontWeight": 5,
            "letterSpacing": 0,
            "lineHeight": 14,
            "marginLeft": 118.5,
          },
          "type": "Text",
        },
      ],
      "frame": Object {
        "height": 404,
        "width": 327,
        "x": 24,
        "y": 452,
      },
      "points": Array [
        Object {
          "x": 24,
          "y": 452,
        },
        Object {
          "x": 351,
          "y": 452,
        },
        Object {
          "x": 351,
          "y": 856,
        },
        Object {
          "x": 24,
          "y": 856,
        },
        Object {
          "x": 187.5,
          "y": 654,
        },
      ],
      "style": Object {
        "alignItems": "center",
        "flexDirection": "column",
        "justifyContent": "flex-start",
        "paddingTop": 0,
      },
      "type": "Block",
    },
  ],
  "frame": Object {
    "height": 452,
    "width": 351,
    "x": 12,
    "y": 404,
  },
  "points": Array [
    Object {
      "x": 12,
      "y": 404,
    },
    Object {
      "x": 363,
      "y": 404,
    },
    Object {
      "x": 363,
      "y": 856,
    },
    Object {
      "x": 12,
      "y": 856,
    },
    Object {
      "x": 187.5,
      "y": 630,
    },
  ],
  "style": Object {
    "alignItems": "center",
    "flexDirection": "column",
    "justifyContent": "flex-start",
    "paddingTop": 16,
  },
  "type": "Block",
}
`;
